### ISSUE TITLE
Get ISV payment failure reason

### ISSUE THREAD

**Issue by fmarchalemisys** [VivaPayments/API#2338]
Hi,

I'm starting an ISV payment with [`/checkout/v2/isv/orders?merchantId=`](https://developer.vivawallet.com/isv-partner-program/payment-isv-api/#tag/Payments/paths/~1checkout~1v2~1isv~1orders/post). It returns an order code and nothing else.

When the payment fails (with visa test card 4147 4630 1111 0141, for instance), the customer is redirected to the failure url  configured for the website. The query part of the url contains the order code and an `eventId`. The event ID may easily be altered by a malicious party. So, I would like to double check it with the api.

The endpoint to get the [ISV transaction by order code](https://developer.vivawallet.com/isv-partner-program/payment-isv-api/#tag/Retrieve-Transactions/paths/~1api~1transactions~1?ordercode={ordercode}/get) returns an empty transaction list. So, it can't be used to check the transaction error status.

Fortunately, the [non ISV endpoint](https://developer.vivawallet.com/apis-for-payments/payment-api/#tag/Payments/paths/~1api~1orders~1{orderCode}/get) does return the status but it feels weird to call the non-ISV api to get the ISV transaction status and the status is nowhere as detailed as the `eventId` of the url.

Is there an api endpoint to reliably get the payment status such as the [`eventId`](https://developer.vivawallet.com/integration-reference/response-codes/#event-id-codes) untrustily passed in the url's query when the user returns to the website?

**Comment by VivaWalletCustomerSupport** [comment]
Hello @fmarchalemisys,

Thank you very much for contacting us. We would like to suggest you to use our [Transaction failed webhook](https://developer.vivawallet.com/webhooks-for-payments/transaction-failed/). This webhook will be sent when a failed customer payment has been attempted to one of your wallets. The reason the payment failed will be indicated by the [‘ResponseEventId’ code](https://developer.vivawallet.com/integration-reference/response-codes/#event-id-codes). 

Kind regards,
Viva Wallet E-commerce Support Team

**Comment by fmarchalemisys** [comment]
Hi,

Thanks for the suggestion.

Unless I'm mistaken, webhooks are configured on each merchant's interface individually. It is not possible to configure them on the main ISV account nor for all the merchants at once. Therefore, I sadly expect some merchants will start accepting payments long before the missing webhook is noticed :unamused: 

Even if the webhook is correctly configured, there is no guarantee the message is received and processed before the user returns to the website. I don't want to show the customer a message such as "It seems you completed your payment. If it is indeed the case, you'll eventually receive the purchased entrance ticket" :laughing: 

So, I guess, for now, my best option is to use the non-ISV api call, right?

**Comment by VivaWalletCustomerSupport** [comment]
Hello @fmarchalemisys,

Thank you for your prompt response. Can you please send us your Retrieve Transaction by Order Code API request and the response you are getting? Thank you in advance.

Kind regards,
Viva Wallet E-commerce Support Team

**Comment by fmarchalemisys** [comment]
Hi,

Thanks for your super fast answer.

The GET query is sent to `https://demo.vivapayments.com/api/transactions/?ordercode=1967007173057608` with the header `Authorization: Basic N…==`. The merchant ID and its api key compose the base64 encoded authorization (the OAuth token or the ISV client ID and secret returns a 401 HTTP status).

The response contains a json payload:

```
{
  "Transactions": [],
  "ErrorCode": 0,
  "ErrorText": null,
  "TimeStamp": "2023-02-22T14:00:28.0462829+02:00",
  "CorrelationId": null,
  "EventId": 0,
  "Success": true
}
```

The response's header is:

```
{
  "content-type": [
    "application/json; charset=utf-8"
  ],
  "content-length": [
    "146"
  ],
  "x-viva-correlationid": [
    "23-053-6913B14D"
  ],
  "rp": [
    "03"
  ],
  "date": [
    "Wed, 22 Feb 2023 12:00:28 GMT"
  ],
  "server-timing": [
    "cdn-cache; desc=MISS",
    "edge; dur=104",
    "origin; dur=21",
    "ak_p; desc=\"465852_34718829_996462252_12552_7471_14_0\";dur=1"
  ],
  "strict-transport-security": [
    "max-age=31536000 ; includeSubDomains",
    "max-age=31536000;IncludeSubDomains"
  ],
  "set-cookie": [
    "ak_bmsc=5F9AD…; Domain=.vivapayments.com; Path=/; Expires=Wed, 22 Feb 2023 14:00:27 GMT; Max-Age=7199; Secure; HttpOnly"
  ]
}
```

The cookie is redacted because I don't know if it contains sensitive information.

**Comment by adoyen** [comment]
Hello,

I have the same issue while trying to retrieve information using ISV credentials (get transaction by order code).
It always end up to error 401 : "Message":"Authorization has been denied for this request." while doing, as asked, the base64 encoding of ISVID:MerchantId:ISVSecret in basic Authorization (see my test 'x-viva-correlationid': '23-054-8B0E5D02')

Maybe making it work correctly will fix the issue.


Concerning the issue with the webhook. I agree with [fmarchalemisys](https://github.com/fmarchalemisys), it will be way better to be able to specify webhooks url and redirection url while creating the order (so we can add merchantId or other useful information in the url) or at least, in the ISV account.

**Comment by fmarchalemisys** [comment]
Hi @adoyen ,

I had http error code 401 too until I started using the basic authentication (not OAuth) with the merchant's ID and api key.

It's a huge inconsistency of the api. That endpoint doesn't accept the ISV's ID and secret.

**Comment by adoyen** [comment]
Hello @fmarchalemisys,
On my side, for the moment, I use field merchantTrns to specify an internal transaction id.
And then, when receiving the webhook, I use the orderCode to retrieve my internal transaction and the transactionId transmitted in the webhook to retrieve the complete transaction (using isv oauth).

After that, I check the merchantTrns and the orderCode to ensure it's correct.
But indeed, it would be better to have the transactionId directly after creation of the order (using the  Retrieve Transaction by Order Code API).

**Comment by VivaWalletCustomerSupport** [comment]
Hello @fmarchalemisys, @adoyen 

Thank you both for your swift responses. We are working on the issue, we will let you know as soon as we have any update.

Kind regards,
Viva Wallet E-commerce Support Team

**Comment by VivaWalletCustomerSupport** [comment]
Hello @fmarchalemisys, @adoyen,

Thank you for waiting. Can you please clarify which credentials you are using from your ISV account to execute the Retrieve Transaction by Order Code API request? 

Kind regards,
Viva Wallet E-commerce Support Team

**Comment by adoyen** [comment]
Hello @VivaWalletCustomerSupport ,

ISVId and ISVSecret :
![image](https://user-images.githubusercontent.com/1709103/220906991-a294d5d1-d4c1-45ea-84dc-6f7cc102d081.png)

Then, merchantId:
![image](https://user-images.githubusercontent.com/1709103/220907184-87025024-5e85-42ba-9286-45ed01deff00.png)

So the Basic authentication is Base64(isvID:merchantId:isvSecret

Here is the part of my code:

> let auth = `${Meteor.settings.vivawallet.isv.id}:${merchantId}:${Meteor.settings.vivawallet.isv.secret}`
>  
>   const callOptions = {
>     headers: {
>       Authorization: "Basic " + Buffer.from(auth).toString("base64")
>     }}
> 

The ISVId and ISVsecret are working well for oauth2 authentication.
The MerchantId works well while creating an order.

I tried in postman either:
![image](https://user-images.githubusercontent.com/1709103/220911592-6ebea2b5-f824-4094-91d5-458471fadd3a.png)


And it's also not working.

**Comment by fmarchalemisys** [comment]
In my case, I use the Authorization header with this content (php):

```php
sprintf("Basic %s", base64_encode(sprintf("%s:%s", $merchantId, $apiKey)));
```

Where `$merchantId` and `$apiKey` are those values from the merchant's account:
![image](https://user-images.githubusercontent.com/30406081/220924085-4d3790a5-b45b-4cde-92f4-3b4c353880b6.png)

**Comment by fmarchalemisys** [comment]
@VivaWalletCustomerSupport to be clear; although I currently access the endpoint with the merchant's ID and api key, it would be more convenient and logical to access the endpoint with the ISV's credentials.

The endpoint to [get the transaction by transaction ID](https://developer.vivawallet.com/isv-partner-program/payment-isv-api/#tag/Retrieve-Transactions/paths/~1checkout~1v2~1isv~1transactions~1{transactionId}?merchantId={merchantId}/get) does accept an OAuth token obtained with the ISV's ID and secret. Accepting the same token to get the transactions by order code would be more consistent.

**Comment by VivaWalletCustomerSupport** [comment]
Hello @fmarchalemisys, @adoyen,

Thank you for waiting. We would like to inform you that our [Retrieve Transaction by Order Code API request](https://developer.vivawallet.com/isv-partner-program/payment-isv-api/#tag/Retrieve-Transactions/paths/~1api~1transactions~1?ordercode=%7Bordercode%7D/get) requires a set of credentials that are provided by us. We kindly suggest you to contact your sales representative and ask the credentials for the above request. We apologize for the inconvenience.

Kind regards,
Viva Wallet E-commerce Support Team

**Comment by fmarchalemisys** [comment]
@VivaWalletCustomerSupport  are you talking about the [ISV program](https://www.vivawallet.com/isv-program)?

I have such an access and @adoyen mentioned the ISV ID and secret in his message :thinking:

**Comment by fmarchalemisys** [comment]
We went a bit off track. Let me rephrase my question.

Creating an [ISV online payment](https://developer.vivawallet.com/isv-partner-program/payment-isv-api/#tag/Payments/paths/~1checkout~1v2~1isv~1orders/post) returns an order code.

When the customer returns to the website after a payment failure, the url contains the order code and an event ID.

The webhook might not be configured or might take some time to arrive and be processed.

So, when the user is redirected to our website after a failed payment, all we have is the order code and the event ID.

It is standard procedure to not trust a value the user could have altered. Therefore, I would like to query the api to know, for sure, the payment status.

Having only an order code, the only obvious choice is to call [get transaction by order code](https://developer.vivawallet.com/isv-partner-program/payment-isv-api/#tag/Retrieve-Transactions/paths/~1api~1transactions~1?ordercode={ordercode}/get). I'm an ISV partner. Calling it should not be a problem.

Alas, in case of failure, the above api endpoint returns an empty transaction list. I can't log any meaningful error message to help our support in investigating the issue. The failed payment doesn't show in the VW web interface either.

Aside from the fact that the `get transaction by order code` requires a merchant ID and api key instead of the expected ISV ID and secret, the main point is that I can't show any meaningful error message except the terse "something went wrong but we don't know what".

In short, I'm looking for a way of getting a meaningful error message from a trusted source in case of failed payment :slightly_smiling_face:

**Comment by VivaWalletCustomerSupport** [comment]
Hello @fmarchalemisysm,

Thank you for waiting. If wrong credentials have been used when calling the [get transaction by order code](https://developer.vivawallet.com/isv-partner-program/payment-isv-api/#tag/Retrieve-Transactions/paths/~1api~1transactions~1?ordercode=%7Bordercode%7D/get) API request, regardless of if the order code has been paid or not, the response will be the 2 following scenarios:

1. 401 Unauthorized.
2. The parameters of the response will have the value null. 

Can you please execute the request with one paid and one failed order code and send us the responses you are getting?

The most effective solution is webhooks. Our webhooks have a retry policy to find the endpoint for 3 days.

Kind regards,
Viva Wallet E-commerce Support Team

**Comment by adoyen** [comment]
Hello @VivaWalletCustomerSupport ,

It's indeed what we are saying. Getting the transaction by transaction ID is working as expected, using oauth token:
![image](https://user-images.githubusercontent.com/1709103/221218964-23ead464-4c36-495f-9d0e-4b49ca86af7c.png)

But, getting the same transaction by orderCode like this (with the same variables) gives us a 401 unauthorized while it shouldn't:
![image](https://user-images.githubusercontent.com/1709103/221219235-335cef97-7009-4c0f-8a1c-8e7d327f52d1.png)

And, as @fmarchalemisys says, we should be able to retrieve transaction by orderCode using the same authentication process (oauth2 and token) like in retrieve transaction by id.

So why our credentials are not working ? (Pay attention in the second screenshot how I concatenate the login field and how the authentication is done. Maybe the url to use is not correct and should contain "isv" somewhere ?).
And why isn't it oauth2 everywhere ?

Kind regards

**Comment by VivaWalletCustomerSupport** [comment]
Hello @adoyen, @fmarchalemisys,

Thank you for the response. As I mentioned above this occurs because wrong credentials have been used on the API request. This API request requires a set of credentials that cannot be found on your ISV account. We should provide these credential.s Please contact your sales representative and ask for those credentials.

Kind regards,
Viva Wallet E-commerce Support Team

**Comment by fmarchalemisys** [comment]
Hi @VivaWalletCustomerSupport ,

You mean there are **two** sets of credentials to access the api, depending on the endpoint??? :open_mouth: 

I'm curious. Would you mind sharing the rationale behind that design decision?

**Comment by adoyen** [comment]
Hi @VivaWalletCustomerSupport ,
My sale representative told me he hasn't any other credentials for ISV account.
I will ask him again but I think I'm going in an infinite loop...

**Comment by adoyen** [comment]
Hi @VivaWalletCustomerSupport,
My sale representative told me to use the merchant id / secret, like @fmarchalemisys did.
I'm not convinced to do that, so I go further in de development since I'm thinking this api, as it is, will be probably of no use.
I would prefer to use webhook.

Hi @fmarchalemisys,
So I tested also this api (get transaction by order code) using merchant credentials and I have the same issue at the beginning : transactions is an empty array.
It seems it's filled in after paying or canceling the transaction.

I tried to validate the payment and the transaction appeared in the "Transactions" array.

**Comment by fmarchalemisys** [comment]
Hi @adoyen ,

@VivaWalletCustomerSupport  told me, on issue #2345 , that transactions are simply dropped when the customer fails to complete 3DS check. The only trace you'll ever get about transactions not completing 3DS is the event ID when the user is redirected to the website (provided such a redirection is correctly setup in the merchant's account).

If the user completes 3DS but the transaction fails for [any of those reasons](https://developer.vivawallet.com/integration-reference/test-cards-and-environments/#trigger-a-decline), a webhook is received but I haven't checked yet if the transaction list contains the failed transaction or if it is still empty.

On my side, as long as the merchant's ID and api key must be used to [get the transaction using the ISV api call](https://developer.vivawallet.com/isv-partner-program/payment-isv-api/#tag/Retrieve-Transactions/paths/~1api~1transactions~1?ordercode=%7Bordercode%7D/get) and the response doesn't contain the information I need, I choose to use the [regular api call](https://developer.vivawallet.com/apis-for-payments/payment-api/#tag/Payments/paths/~1api~1orders~1%7BorderCode%7D/get). That endpoint returns a limited status but it works even even when 3DS did not complete. I haven't found any drawback, yet :sweat_smile:

**Comment by VivaWalletCustomerSupport** [comment]
Hello @adoyen, @fmarchalemisys,

We are sorry for the late response. The solution we would like to recommend is the Transaction failed Webhook. In any case can you please send us your ISV ID or your Merchant ID to investigate the issue of retrieving transaction by order code API request? Thank you in advance.

Kind regards,
Viva Wallet E-commerce Support Team

**Comment by VivaWalletCustomerSupport** [comment]
Hello @fmarchalemisys, @adoyen,

We just wanted to follow up on the reply we sent. We were simply wondering if you read our response.

In any case this ticket will remain active for at least 24 hours, so that you can come back if you wish.

However, if you have any questions, please don't hesitate to reply here and we'd be happy to help!

Kind regards,
Viva Wallet E-commerce Support Team

**Comment by adoyen** [comment]
Hello @VivaWalletCustomerSupport, my demo ISV ID is t5g2l17g9rxgdvk0cksjp2xq98q8e2342j454mx42hr38.apps.vivapayments.com

**Comment by fmarchalemisys** [comment]
Hi @VivaWalletCustomerSupport ,

Sorry for the delay. I had to switch to other urgent tasks.

The ISV ID is `t5nh16oi37cdq24tj05khib53lrhoc8340w4z7g21uc86.apps.vivapayments.com`
The merchant's ID is `5f613870-d7fa-4293-a486-1b1005b7ef43`.

**Comment by VivaWalletCustomerSupport** [comment]
Hello @fmarchalemisys, @adoyen,

Thank you for the additional information. We kindly suggest sending us your e-mail address so we can provide you the necessary credentials for the [retrieve transaction by order code](https://developer.vivawallet.com/isv-partner-program/payment-isv-api/#tag/Retrieve-Transactions/paths/~1api~1transactions~1?ordercode=%7Bordercode%7D/get) API request. Thank you in advance.

Kind regards,
Viva Wallet E-commerce Support Team

**Comment by fmarchalemisys** [comment]
Hi @VivaWalletCustomerSupport ,

Thanks for your assistance.

I do not wish to share my email address on a public forum (I receive my share of spams as it is :slightly_smiling_face: ).

So, I'll settle for the solution I have now. I query a simplified status using the merchant's ID and secret using the [normal api](https://developer.vivawallet.com/apis-for-payments/payment-api/#tag/Payments/paths/~1api~1orders~1%7BorderCode%7D/get).

It's far from perfect but it's no less cumbersome than using yet another set of ultra confidential credentials to query an api.

**Comment by adoyen** [comment]
Hello @VivaWalletCustomerSupport,
I think the same about the email and yet another set of credentials.
I'll use webhooks, retrieve transaction by id and auto expire on my app side.

If you really want to fix your api, please, let us use oauth token everywhere for ISV and avoid having 3 different domains as endpoints for your different api calls.

Furthermore, it would be interesting you allow us to send webhook endpoints and url redirection while creating the order instead of having to explain our clients how to configure these for us on their merchant account.

Best regards.

**Comment by VivaWalletCustomerSupport** [comment]
Hello @adoyen, @fmarchalemisys,

Thank you for the feedback. We have noted your comments for future improvements to our API's. 

In any case this ticket will remain active for at least 24 hours, so that you can come back if you wish.

However, if you have any questions, please don't hesitate to reply here and we'd be happy to help!

Kind regards,
Viva Wallet E-commerce Support Team

**Comment by VivaWalletCustomerSupport** [comment]
Hello @adoyen, @fmarchalemisys,

This ticket will close for now. Please don't hesitate to raise a new ticket if you need further assistance.

Kind regards,
Viva Wallet E-commerce Support Team
